File: core/planner/src/main/java/ch/epfl/dias/calcite/adapter/pelago/metadata/PelagoRelMdPercentageOriginalRows.java
package ch.epfl.dias.calcite.adapter.pelago.metadata;

import org.apache.calcite.rel.metadata.BuiltInMetadata;
import org.apache.calcite.rel.metadata.ChainedRelMetadataProvider;
import org.apache.calcite.rel.metadata.MetadataDef;
import org.apache.calcite.rel.metadata.MetadataHandler;
import org.apache.calcite.rel.metadata.ReflectiveRelMetadataProvider;
import org.apache.calcite.rel.metadata.RelMdPercentageOriginalRows;
import org.apache.calcite.rel.metadata.RelMetadataProvider;
import org.apache.calcite.rel.metadata.RelMetadataQuery;
import org.apache.calcite.util.BuiltInMethod;

import com.google.common.collect.ImmutableList;

import ch.epfl.dias.calcite.adapter.pelago.PelagoDeviceCross;
import ch.epfl.dias.calcite.adapter.pelago.PelagoPack;
import ch.epfl.dias.calcite.adapter.pelago.PelagoRouter;
import ch.epfl.dias.calcite.adapter.pelago.PelagoUnpack;

public class PelagoRelMdPercentageOriginalRows
    implements MetadataHandler<BuiltInMetadata.PercentageOriginalRows> {
  private static final PelagoRelMdPercentageOriginalRows INSTANCE =
      new PelagoRelMdPercentageOriginalRows();

  public static final RelMetadataProvider SOURCE =
      ChainedRelMetadataProvider.of(
          ImmutableList.of(
              ReflectiveRelMetadataProvider.reflectiveSource(
                  BuiltInMethod.PERCENTAGE_ORIGINAL_ROWS.method, PelagoRelMdPercentageOriginalRows.INSTANCE),
              RelMdPercentageOriginalRows.SOURCE));

  //~ Methods ----------------------------------------------------------------

  private PelagoRelMdPercentageOriginalRows() {}

  public MetadataDef<BuiltInMetadata.PercentageOriginalRows> getDef() {
    return BuiltInMetadata.PercentageOriginalRows.DEF;
  }

  public Double getPercentageOriginalRows(PelagoUnpack rel, RelMetadataQuery mq) {
    return mq.getPercentageOriginalRows(rel.getInput());
  }

  public Double getPercentageOriginalRows(PelagoPack rel, RelMetadataQuery mq) {
    return mq.getPercentageOriginalRows(rel.getInput());
  }

  public Double getPercentageOriginalRows(PelagoRouter rel, RelMetadataQuery mq) {
    return mq.getPercentageOriginalRows(rel.getInput());
  }

  public Double getPercentageOriginalRows(PelagoDeviceCross rel, RelMetadataQuery mq) {
    return mq.getPercentageOriginalRows(rel.getInput());
  }
}

// End RelMdPercentageOriginalRows.java

File: core/planner/src/main/scala/ch/epfl/dias/calcite/adapter/pelago/metadata/PelagoRelMdPercentageOriginalRows.scala
/*
    Proteus -- High-performance query processing on heterogeneous hardware.

                            Copyright (c) 2017
        Data Intensive Applications and Systems Laboratory (DIAS)
                École Polytechnique Fédérale de Lausanne

                            All Rights Reserved.

    Permission to use, copy, modify and distribute this software and
    its documentation is hereby granted, provided that both the
    copyright notice and this permission notice appear in all copies of
    the software, derivative works or modified versions, and any
    portions thereof, and that both notices appear in supporting
    documentation.

    This code is distributed in the hope that it will be useful, but
    WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. THE AUTHORS
    DISCLAIM ANY LIABILITY OF ANY KIND FOR ANY DAMAGES WHATSOEVER
    RESULTING FROM THE USE OF THIS SOFTWARE.
 */

package ch.epfl.dias.calcite.adapter.pelago.metadata

import ch.epfl.dias.calcite.adapter.pelago.rel.{
  PelagoDeviceCross,
  PelagoPack,
  PelagoRouter,
  PelagoUnpack
}
import com.google.common.collect.ImmutableList
import org.apache.calcite.rel.metadata._
import org.apache.calcite.util.BuiltInMethod

import java.lang

object PelagoRelMdPercentageOriginalRows {
  private val INSTANCE = new PelagoRelMdPercentageOriginalRows
  val SOURCE: RelMetadataProvider = ChainedRelMetadataProvider.of(
    ImmutableList.of(
      ReflectiveRelMetadataProvider.reflectiveSource(
        BuiltInMethod.PERCENTAGE_ORIGINAL_ROWS.method,
        PelagoRelMdPercentageOriginalRows.INSTANCE
      ),
      RelMdPercentageOriginalRows.SOURCE
    )
  )
}

class PelagoRelMdPercentageOriginalRows private ()
    extends MetadataHandler[BuiltInMetadata.PercentageOriginalRows] {
//~ Methods ----------------------------------------------------------------
  override def getDef: MetadataDef[BuiltInMetadata.PercentageOriginalRows] =
    BuiltInMetadata.PercentageOriginalRows.DEF

  def getPercentageOriginalRows(
      rel: PelagoUnpack,
      mq: RelMetadataQuery
  ): lang.Double =
    mq.getPercentageOriginalRows(rel.getInput)

  def getPercentageOriginalRows(
      rel: PelagoPack,
      mq: RelMetadataQuery
  ): lang.Double =
    mq.getPercentageOriginalRows(rel.getInput)

  def getPercentageOriginalRows(
      rel: PelagoRouter,
      mq: RelMetadataQuery
  ): lang.Double =
    mq.getPercentageOriginalRows(rel.getInput)

  def getPercentageOriginalRows(
      rel: PelagoDeviceCross,
      mq: RelMetadataQuery
  ): lang.Double =
    mq.getPercentageOriginalRows(rel.getInput)
}
